### DISCUSSION TITLE
Could you add some quantum algorithm examples?

### DISCUSSION THREAD

**Discussion by haiyongsong1921** [NVIDIA/cuQuantum#15] [Ideas]
It will be great to add some practical examples, like grover search.

**Comment by tlubowe** [comment]
This is a great idea. We have some examples in the samples, and python/samples folder but nothing like grover search. 

We have been thinking about how to improve user's experience getting started and this could be a great way to help with that. 

Thank you for the input!

**Comment by haiyongsong1921** [comment]
No problem. I am very excited and looking forward to your output

**Reply by tlubowe** [reply]
In the meantime, if you are using cirq, pennylane or qiskit with cuQuantum you can leverage their examples as well. We are an accelerator library which just speeds up popular simulators, many of these frameworks have examples in which changing the backend would effectively enable this.

**Reply by haiyongsong1921** [reply]
Yes, it's an option. But as a C/C++ developer, I preferred to use cuStateVec/cuTensorNet directly, not the python wrapper, so instead of the UT like in examples folder, it would be better to add some practical examples.

**Reply by tlubowe** [reply]
May I ask if you are building a simulator?  We alone are not a simulator, so we could in theory hard code something, which shows how a simulator developer might leverage multiple APIs in cuQuantum.

**Reply by haiyongsong1921** [reply]
Yes, I am trying to build a simulator with the cuQuantum C lib, just for learn and research for Quantum computing.  Do you mean it is impossible for me to build a simulator with cuQuantum?

**Reply by leofang** [reply]
Hi @haiyongsong1921 sorry we dropped the ball, I was just going through the open discussions and found this thread. 
> Do you mean it is impossible for me to build a simulator with cuQuantum?

No, this is definitely not what we meant. It is totally possible, and is what our QC framework partners have done, to build a quantum simulator *on top of* cuQuantum libraries. However, as Tom said we are not a _simulator_, what we currently provide are GPU-accelerated, *low-level* APIs that can be wrapped to build a _simulator backend_. In order to deliver examples like Grover search that you asked for, it could require thousands if not hundreds lines of C/C++ code, which is impractical for the demo purpose, and we consider them to be better addressed with  *high-level* APIs, such as our [`CircuitToEinsum`](https://docs.nvidia.com/cuda/cuquantum/python/api/generated/cuquantum.CircuitToEinsum.html) that can turn a quantum circuit to a tensor network for cuTensorNet to take over.

If you already have built your simulator in C/C++ with appropriate abstraction, we are happy to discuss with you how some of your functions could be replaced by our APIs so as to accelerate them with GPUs. But, like Tom said, we usually point users to one of the established projects -- such as Qiskit Aer or Qsimcirq -- that have integrated cuQuantum, so that people don't have to start from scratch.

Let us know if this answers your question, and please start a new discussion thread for further questions. Thank you.

**Reply by haiyongsong1921** [reply]
Hi @leofang @tlubowe , thanks a lot. I will start a new discussion for the next question.

**Comment by mtjrider** [comment]
@haiyongsong1921 
I am closing this discussion due to inactivity and a lack of a specified answer.
If you have additional follow-up, please post another discussion/issue referencing this one.
